kvfollowerreadsccl: implement txn retries for nearest_only=True

otan opened the pull request:
See commit for details.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68969)
<!-- Reviewable:end -->

otan commented:
i'm a little iffy at the schema changes, so would appreciate a look from @cockroachdb/sql-schema on that.

otan commented:
also @rytaft if you could look at the opt changes and let me know what you think about whether we can do better on the AOST matching clause here if outer != in (see the test changed needed in as_of to make this happen)

ajwerner reviewed (commented):
Is there a design for how this is supposed to work regarding schema descriptor timestamps and bounded staleness time ranges?


Historical descriptors have a stricter safety policy than current descriptors for which you actively hold a lease. To use historical descriptors, you must use them during their durable validity interval (between their ModificationTime and the successor descriptor's Modification time, exclusive). This is because we don't track the moment when we've stopped assuming that the predecessor version might be active. 

Can you help me understand some timestamps in play here? My assumption is that the `ReadTimestamp` on the `Txn` is set to the minimum timestamp that the query is allowed to see (i.e. the value computed from the AOST clause). Let's think about this in terms of a timeline. Imagine user specifies that their query should use a timestamp no older than `a` and now is `g`.

Now we need to plan the query with some descriptor versions. Imagine our query uses two descriptors and those descriptors have gone through some schema changes during the time interval. Namely `d1` saw a change at `c` and at `e` and `d2` saw changes at `b`, `d`, and `f`. 


```
--time-->
abc  d   e   f  g // timestamp labels
[...............] // interval for bounded staleness query, g == now
..|......|....... // version validity intervals for two versions of descriptor d1
.|...|.......|... // version validity intervals for two versions of descriptor d2
```

So, there should exist the following distinct plans as far as the optimizer is concerned:
`[a, b)`, `[b, c)`, `[c, d)`, `[d, e)`, `[e, f)`, and `[f, g)`. 


When I look back at the RFC, there's some language about using follower reads and a similar lookup protocol to find descriptors and then to try to make that cohere [here](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20210519_bounded_staleness_reads.md#table-schemas). I don't think I ever internalized this part of the proposal at all and assumed it would just use leases and other cached historical descriptors given we rely on [system tables remaining available](https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20210519_bounded_staleness_reads.md#system-range-unavailability).

So, in my idealized world, in the face of a schema change, we'd 
1) Plan at the present time
2) Look at the descriptors used to plan and bound the minimum timestamp to the earliest modification time of any descriptors used. 

In this case it would mean that we'd end up changing the minimum timestamp from `a` to `f` for the purpose of execution and we'd use `g` for planning. 

I may be misunderstanding things and am more than happy to get on calls.  

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-:-MhETbOt8x3Y1HON5zMh:brxxijj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @rytaft)



<!-- Sent from Reviewable.io -->

otan commented:
> When I look back at the RFC, there's some language about using follower reads and a similar lookup protocol to find descriptors and then to try to make that cohere here. I don't think I ever internalized this part of the proposal at all and assumed it would just use leases and other cached historical descriptors given we rely on system tables remaining available.

Ah, yeah... that's because this *used* to that we would give up on looking at the past due to the "schema unavailability" problem, but we changed it to what you saw when we thought we could change this: https://github.com/cockroachdb/cockroach/pull/67837

I probably should have circulated the RFC change when we did this.

> Can you help me understand some timestamps in play here?

I think the pseudocode made sense for conceptualising what was happening, but in actuality (aka when I made this work), we do something like this:

```
a.....b.....c    
|.....|.....
```

Assume (a) a bounded staleness bound, (b) is a schema modification, (c) is "current time".

* The code right now will attempt a staleness read between `[a, c]`, but grabs the latest schema and realises it can only service `[b, c]`. It attempts a read, and if nearest_only=True and the follower is not past `b` it will error.
* We will try, but set the cap at b. The bounded staleness read will attempt to read at `[a, b)`, assuming the schema is read from the point in time `b.Prev()`. The planner side will use the descriptor at time `b.Prev()` to do the planning.
 
This can continue if there are more schema changes, in which case we always do `[max(bounded_staleness_timestamp, schema-modification_time), min(now, point before next schema)]`.

In reference to your question, we do your idealized version first, but then try go backwards as well. Let me know if that doesn't make sense.

ajwerner reviewed (commented):
Thanks, I feel a lot better now. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-:-MhG6XgWCzt1BWzEbiK8:b-txqdax)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @otan, and @rytaft)

___
*[pkg/kv/txn.go, line 1108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhG6BUjCRwgl2w1BjUD:-MhG6BUk8I55tQJfS8yH:b7aj4py) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c381e7ad2c9b3dbd978d4924254ea828fd2d960/pkg/kv/txn.go#L1108)):*
> ```Go
> 			err := errors.AssertionFailedf("%s: ba=%s, txn=%s", s, ba.String(), txn.String())
> 			err = errors.WithContextTags(err, ctx)
> 			panic(err)
> ```

are we happy with this new panic behavior? Is this a lot better than turning it into a roachpb error? It should retain it's assertion failure nature as it goes through that round of boxing and unboxing. 
___
*[pkg/sql/conn_executor.go, line 2191 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhG7cid0loQGZv9ijy_:-MhG7cid0loQGZv9ijya:bjiprr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9906923dd7c3b31ddfcbe16f7c1af868d06d405/pkg/sql/conn_executor.go#L2191)):*
> ```
> 	var minTSErr *roachpb.MinTimestampBoundUnsatisfiableError
> 	if errors.As(err, &minTSErr) {
> ```
total nit: you can scope this variable to ths conditional with:

```
if minTSErr := (*roachpb.MinTimestampBoundUnsatisfiableError)(nil); errors.As(err, &minTSErr) {
    // ...
```
___
*[pkg/sql/conn_executor.go, line 2198 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhG7ziA-zLuj5luw1DL:-MhG7ziA-zLuj5luw1DM:b-xysd0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9906923dd7c3b31ddfcbe16f7c1af868d06d405/pkg/sql/conn_executor.go#L2198)):*
> ```Go
> 				// a minimum timestamp during a bounded staleness read to be greater
> 				// than the maximum staleness bound we put up.
> 				err = errors.Wrapf(
> ```

assertion failure?
___
*[pkg/sql/catalog/descs/collection.go, line 109 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhG8xJc3_7QcM5ncJ76:-MhG8xJc3_7QcM5ncJ77:bi3z9gu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9906923dd7c3b31ddfcbe16f7c1af868d06d405/pkg/sql/catalog/descs/collection.go#L109)):*
> ```Go
> 	// nearest_only=True, in which we want a schema read that should be
> 	// no older than maxTimestampBound.
> 	maxTimestampBound *hlc.Timestamp
> ```

Assuming we need all of this, thoughts on just sticking a `maxTimestampBoundDeadlineHolder` here and then detecting whether that thing has a zero timestamp as opposed to allocating a new one on each use?
___
*[pkg/sql/catalog/descs/collection.go, line 122 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhG9Wem9j2x6NrWeUcW:-MhG9Wem9j2x6NrWeUcX:b-ck18c9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9906923dd7c3b31ddfcbe16f7c1af868d06d405/pkg/sql/catalog/descs/collection.go#L122)):*
> ```Go
> 
> // SetMaxTimestampBound sets the maximum timestamp to read schemas at.
> func (tc *Collection) SetMaxTimestampBound(maxTimestampBound *hlc.Timestamp) {
> ```

I'm still unclear on why we need this. Reasons seem to be:

1) we want to avoid actually using the transaction to go read from the store for the descriptor in the case we don't have a lease
2) we want to sidestep the UpdateDeadline call
3) we want to use this timestamp as opposed to the transaction's read timestamp

Taking them backwards, what is the `ReadTimestamp` on one of these transactions? What does happen on `UpdateDeadline` on one of these transactions? Can you tell that you have one of these transactions just from the `*kv.Txn` object?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68969)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @nvanbenschoten, and @rytaft)

___
*[pkg/kv/txn.go, line 1108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhG6BUjCRwgl2w1BjUD:-MhGAT0k0euaHkdFiF1R:bav1bnv) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c381e7ad2c9b3dbd978d4924254ea828fd2d960/pkg/kv/txn.go#L1108)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

are we happy with this new panic behavior? Is this a lot better than turning it into a roachpb error? It should retain it's assertion failure nature as it goes through that round of boxing and unboxing. 
</blockquote></details>

cc @nvanbenschoten 
___
*[pkg/sql/conn_executor.go, line 2191 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhG7cid0loQGZv9ijy_:-MhGFnV_EVOlMbN6KsjQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9906923dd7c3b31ddfcbe16f7c1af868d06d405/pkg/sql/conn_executor.go#L2191)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	var minTSErr *roachpb.MinTimestampBoundUnsatisfiableError
> 	if errors.As(err, &minTSErr) {
> ```
total nit: you can scope this variable to ths conditional with:

```
if minTSErr := (*roachpb.MinTimestampBoundUnsatisfiableError)(nil); errors.As(err, &minTSErr) {
    // ...
```
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descs/collection.go, line 122 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhG9Wem9j2x6NrWeUcW:-MhGAUguA4GgAZNmBcKq:b-ydylji) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9906923dd7c3b31ddfcbe16f7c1af868d06d405/pkg/sql/catalog/descs/collection.go#L122)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm still unclear on why we need this. Reasons seem to be:

1) we want to avoid actually using the transaction to go read from the store for the descriptor in the case we don't have a lease
2) we want to sidestep the UpdateDeadline call
3) we want to use this timestamp as opposed to the transaction's read timestamp

Taking them backwards, what is the `ReadTimestamp` on one of these transactions? What does happen on `UpdateDeadline` on one of these transactions? Can you tell that you have one of these transactions just from the `*kv.Txn` object?
</blockquote></details>

I was told by @nvanbenschoten to not put things on `kv.Txn`, which is how it ended up here where it is. 
`ReadTimestamp` doesn't change at the moment in a staleness read - should it?  I tried that and a bunch of other things complained based on assertions made that it should not be set during a max staleness read. I've been assuming (based on the APIs given) that it's not important.

Your reasons are correct.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
The non-kv, non-schema stuff <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 18 files at r1, 3 of 3 files at r4, 1 of 2 files at r5, 1 of 19 files at r8, 5 of 10 files at r9, 10 of 13 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-:-MhHAx8lAczxyebOFMes:b-blflqc)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @nvanbenschoten, and @otan)

___
*[pkg/sql/conn_executor.go, line 2409 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhH7fLQ8XtKFsD4NxM9:-MhH7fLQ8XtKFsD4NxMA:b-u6uab2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c9e19d4ef37d0121d42c0586eef0dea47702b51/pkg/sql/conn_executor.go#L2409)):*
> ```Go
> 	// as there was a schema update in between. When retrying, we want to keep the
> 	// same minimum timestamp for the AOST read, but set the maximum timestamp
> 	// to the point just before our failed read to ensure we don't try read data
> ```

nit: try read -> try to read
___
*[pkg/sql/exec_util.go, line 1170 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhH6klh4ch6jt89ifVl:-MhH6klh4ch6jt89ifVm:b-bwx26w) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c9e19d4ef37d0121d42c0586eef0dea47702b51/pkg/sql/exec_util.go#L1170)):*
> ```Go
> 	DistSQLReceiverPushCallbackFactory func(query string) func(rowenc.EncDatumRow, *execinfrapb.ProducerMetadata)
> 
> 	// OnTxnRetry, if sets, will be called if there is a transaction retry.
> ```

nit: if sets -> if set
___
*[pkg/sql/sem/tree/as_of.go, line 93 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhH7HUDBk0cbE2cEa2K:-MhH7HUDBk0cbE2cEa2L:b-wy0irl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c9e19d4ef37d0121d42c0586eef0dea47702b51/pkg/sql/sem/tree/as_of.go#L93)):*
> ```Go
> 	NearestOnly bool
> 	// If this is a bounded staleness read with nearest_only=True, this is set
> 	// when we failed to satisfy a bounded staleness with a nearby replica as
> ```

nit: satisfy a bounded staleness read


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68969)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @nvanbenschoten, and @rytaft)

___
*[pkg/sql/catalog/descs/collection.go, line 122 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhG9Wem9j2x6NrWeUcW:-MhGidb95RFaZVenRJ73:bmy7le7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e9906923dd7c3b31ddfcbe16f7c1af868d06d405/pkg/sql/catalog/descs/collection.go#L122)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I was told by @nvanbenschoten to not put things on `kv.Txn`, which is how it ended up here where it is. 
`ReadTimestamp` doesn't change at the moment in a staleness read - should it?  I tried that and a bunch of other things complained based on assertions made that it should not be set during a max staleness read. I've been assuming (based on the APIs given) that it's not important.

Your reasons are correct.
</blockquote></details>

Sounds good. I'm cool with this. My sincere and heartfelt hope is to soon rework the dependency logic in this here `Collection` object to be such that it is constructed (or reset, I suppose) with a binding to the (abstracted) transaction and we remove `txn` parameter from all of these methods. Seems unlikely to make this release so let's march forward with this and then, when that refactor arrives, we can just set this up at construction time in a more direct manner than MaybeUpdateDeadline. 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 7 of 13 files at r10, 1 of 3 files at r11, 33 of 33 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68969)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @otan)

___
*[pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row, line 145 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKUlSa5sm4_-5Jydaj:-MhKUlSa5sm4_-5Jydak:bselk3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row#L145)):*
> ```
> # successfully. Note the schema changes several times, hence the multiple events
> # output. Note on the first read, event 7 is not a follower read as the query
> # requires reading a descriptor which does not utilize a follower read.
> ```

Can we ignore the descriptor lookup? Or somehow ensure that it's cached at this point? This seems like it will be a source of flakiness. I also don't understand why the descriptor lookup results in a local read, but not a follower read. Have we ensured that the descriptors leaseholder is on n2?
___
*[pkg/sql/conn_executor.go, line 2208 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKXAYt3ixoSxOU0Ndr:-MhKXAYt3ixoSxOU0Nds:b-s4dof1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/conn_executor.go#L2208)):*
> ```Go
> 				)
> 			}
> 			if aost.Timestamp.Less(minTSErr.MinTimestampBound.Prev()) {
> ```

Could you add a comment explaining the `.Prev()` here? It must have something to do with an inclusive/exclusive translation somewhere else, right?
___
*[pkg/sql/conn_executor.go, line 2412 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKXUD16vczNcG5g6le:-MhKXUD16vczNcG5g6lf:be7phfi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/conn_executor.go#L2412)):*
> ```Go
> 	// same minimum timestamp for the AOST read, but set the maximum timestamp
> 	// to the point just before our failed read to ensure we don't try to read
> 	//data which may be after the schema change when we retry.
> ```

nit: missing space at the beginning of this line
___
*[pkg/sql/conn_executor.go, line 2415 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKXsneCvvcvTCiyLeh:-MhKXsneCvvcvTCiyLei:bjqvuel) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/conn_executor.go#L2415)):*
> ```Go
> 	var minTSErr *roachpb.MinTimestampBoundUnsatisfiableError
> 	if err := ex.extraTxnState.autoRetryReason; err != nil && errors.As(err, &minTSErr) {
> 		bound := minTSErr.MinTimestampBound.Prev()
> ```

Would things be easier to reason about it max_timestamp_bound was exclusive instead of inclusive? We can make that change in the KV API if necessary.
___
*[pkg/sql/catalog/descs/descriptor.go, line 212 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKYm6hDpgwPyZE5vP1:-MhKYm6hDpgwPyZE5vP2:bqbrep1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/catalog/descs/descriptor.go#L212)):*
> ```Go
> 		ctx, tc.deadlineHolder(txn), parentID, parentSchemaID, name)
> 	if err != nil {
> ```

nit: stray new line
___
*[pkg/sql/catalog/descs/leased_descriptors.go, line 59 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKZ8IVEIX0BDUSdkwj:-MhKZ8IWAT28UdxqMcS_:b-m90bsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/catalog/descs/leased_descriptors.go#L59)):*
> ```Go
> 	ctx context.Context, deadline hlc.Timestamp,
> ) error {
> 	return nil
> ```

Do we need to use the deadline even for bounded staleness queries? Right now we aren't passing a `MaxTimestampBound` for bounded staleness queries that have not yet retried, but I think this may be faulty. If the bounded staleness read is operating on the current schema, gets delayed, and then chooses a dynamic timestamp beyond the end of the validity period for this schema, won't bad things happen?

cc. @ajwerner 
___
*[pkg/sql/catalog/descs/leased_descriptors.go, line 141 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhK_GGMDdoDqPM6gxSH:-MhK_GGN8YWjfUj2ejRd:b-udeaxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/catalog/descs/leased_descriptors.go#L141)):*
> ```Go
> 		// limitations of AcquireByName for details.
> 		// Note we never should read from store during a bounded staleness read,
> 		// as it is safe to return the schema as non-existent.
> ```

What is the effect of this? Does this mean that if the schema exists but is not cached, we won't observe it and will incorrectly claim that it does not exist?
___
*[pkg/sql/sem/tree/as_of.go, line 96 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKR04_8Hq8eEkeco6t:-MhKR04_8Hq8eEkeco6u:b-rnxkpf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/sem/tree/as_of.go#L96)):*
> ```Go
> 	// have no followers with an up-to-date schema.
> 	// In this case, we want a read that is >= Timestamp and <= MaxTimestampBound.
> 	MaxTimestampBound *hlc.Timestamp
> ```

nit: we generally use the zero-value ot hlc.Timestamp to represent an optional timestamp. There's an `IsEmpty` method you can use to check whether the timestamp is present or not. No reason to incur runtime cost for this.
___
*[pkg/kv/txn.go, line 1108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhG6BUjCRwgl2w1BjUD:-MhKRUPZBDp4OiJ748Hd:bcotdb6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c381e7ad2c9b3dbd978d4924254ea828fd2d960/pkg/kv/txn.go#L1108)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

cc @nvanbenschoten 
</blockquote></details>

We could turn this into an error. Done in https://github.com/cockroachdb/cockroach/pull/69046.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68969)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @nvanbenschoten)

___
*[pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row, line 145 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKUlSa5sm4_-5Jydaj:-MhKtTxo4CudAo9O5slR:b13oacv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/ccl/kvccl/kvfollowerreadsccl/testdata/boundedstaleness/single_row#L145)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we ignore the descriptor lookup? Or somehow ensure that it's cached at this point? This seems like it will be a source of flakiness. I also don't understand why the descriptor lookup results in a local read, but not a follower read. Have we ensured that the descriptors leaseholder is on n2?
</blockquote></details>

it's because `kv.OnlyFollowerReads` ensures *everything* is a follower read, including the descriptor lookup.
i've changed this to wait for an instance of follower read before continuing so we have the cache entry.
___
*[pkg/sql/conn_executor.go, line 2208 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKXAYt3ixoSxOU0Ndr:-MhKtil0-TI4eE7WE5wv:bwr0w9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/conn_executor.go#L2208)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you add a comment explaining the `.Prev()` here? It must have something to do with an inclusive/exclusive translation somewhere else, right?
</blockquote></details>

ah, it's not needed here.
___
*[pkg/sql/conn_executor.go, line 2415 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKXsneCvvcvTCiyLeh:-MhKtyn17YiT6Wwmh71J:boh8emq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/conn_executor.go#L2415)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Would things be easier to reason about it max_timestamp_bound was exclusive instead of inclusive? We can make that change in the KV API if necessary.
</blockquote></details>

still needs the `Prev()` for the schema descriptor lookup, so i prefer it this way tbh.
___
*[pkg/sql/catalog/descs/leased_descriptors.go, line 59 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKZ8IVEIX0BDUSdkwj:-MhL9nUzEMKJbiz9gL-v:bh1iuf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/catalog/descs/leased_descriptors.go#L59)):*
>  If the bounded staleness read is operating on the current schema, gets delayed, and then chooses a dynamic timestamp beyond the end of the validity period for this schema, won't bad things happen?

isn't that ok? we would truncate the min timestamp for that read anyway right, and retry would take care of the rest? unless i'm misunderstanding.
___
*[pkg/sql/catalog/descs/leased_descriptors.go, line 141 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhK_GGMDdoDqPM6gxSH:-MhKubHz2rAjtZxlRPBS:bh97qtr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/catalog/descs/leased_descriptors.go#L141)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What is the effect of this? Does this mean that if the schema exists but is not cached, we won't observe it and will incorrectly claim that it does not exist?
</blockquote></details>

The schema doesn't exist; we're falling back to the KV read in case it does. But for bounded staleness, we should avoid that because that only observes the "latest" read. 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 9 of 9 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68969)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @otan)

___
*[pkg/sql/catalog/descs/leased_descriptors.go, line 59 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKZ8IVEIX0BDUSdkwj:-MhMWoTTAlb0QBVl8wGP:b-qz5lo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/catalog/descs/leased_descriptors.go#L59)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

>  If the bounded staleness read is operating on the current schema, gets delayed, and then chooses a dynamic timestamp beyond the end of the validity period for this schema, won't bad things happen?

isn't that ok? we would truncate the min timestamp for that read anyway right, and retry would take care of the rest? unless i'm misunderstanding.
</blockquote></details>

We would truncate the min timestamp, but what about the max timestamp? I'm concerned that we have a lease on a descriptor from time 20 to time 30 and we have a min_timestamp_bound of 10. We bump the min_timestamp_bound to 20, but what if we end up evaluating the bounded staleness read at time 35? What if a schema change or two landed between 30 and 35? Could that lead to incoherence between the table data and the schema we use to decode that data?

The TL;DR is that I think we may need to always set a max timestamp bound.
___
*[pkg/sql/catalog/descs/leased_descriptors.go, line 141 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhK_GGMDdoDqPM6gxSH:-MhMYDRq6-wT2opwyFo9:b-r395fg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/catalog/descs/leased_descriptors.go#L141)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

The schema doesn't exist; we're falling back to the KV read in case it does. But for bounded staleness, we should avoid that because that only observes the "latest" read. 
</blockquote></details>

I see, this is about dropped/truncated tables, not newly created ones. This makes sense then. Returning that the schema is non-existent is the "least stale" result possible. Makes sense.
___
*[pkg/sql/sem/tree/as_of.go, line 96 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKR04_8Hq8eEkeco6t:-MhMWYXBDmrbDt1vAoH7:bhzuujo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/sem/tree/as_of.go#L96)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: we generally use the zero-value ot hlc.Timestamp to represent an optional timestamp. There's an `IsEmpty` method you can use to check whether the timestamp is present or not. No reason to incur runtime cost for this.
</blockquote></details>

follow-up nit: mention that the value can be set to the zero value when a query has no maximum timestamp bound.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/catalog/descs/leased_descriptors.go, line 59 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKZ8IVEIX0BDUSdkwj:-MhMZ1O24vH31T43cGQY:bfbes28) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/catalog/descs/leased_descriptors.go#L59)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We would truncate the min timestamp, but what about the max timestamp? I'm concerned that we have a lease on a descriptor from time 20 to time 30 and we have a min_timestamp_bound of 10. We bump the min_timestamp_bound to 20, but what if we end up evaluating the bounded staleness read at time 35? What if a schema change or two landed between 30 and 35? Could that lead to incoherence between the table data and the schema we use to decode that data?

The TL;DR is that I think we may need to always set a max timestamp bound.
</blockquote></details>

Not sure I understand, this controls what time the schema should be read.

For the first read, if we have a min_timestamp_bound of 10 in your example, max is not set, so we try read from latest (and this code is not used). If we have something between 30-35, then the min_timestamp_bound will be reset to something in 30-35 (say 30).
If that fails, we have set a min_timestamp_bound of say 30 in the read because of the bump, then max timestamp bound is set to 29.
The next time, we always read the schema from time 29. We then use this deadline holder. Does the deadline matter in this case when we know there is a schema that is later?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68969)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @otan)

___
*[pkg/sql/catalog/descs/leased_descriptors.go, line 59 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKZ8IVEIX0BDUSdkwj:-MhM_wBX1VkXNh7Rvs4J:b-ffa9du) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/catalog/descs/leased_descriptors.go#L59)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Not sure I understand, this controls what time the schema should be read.

For the first read, if we have a min_timestamp_bound of 10 in your example, max is not set, so we try read from latest (and this code is not used). If we have something between 30-35, then the min_timestamp_bound will be reset to something in 30-35 (say 30).
If that fails, we have set a min_timestamp_bound of say 30 in the read because of the bump, then max timestamp bound is set to 29.
The next time, we always read the schema from time 29. We then use this deadline holder. Does the deadline matter in this case when we know there is a schema that is later?
</blockquote></details>

Isn't `UpdateDeadline` used to ensure a transaction does not commit beyond the validity of the schema descriptors that it is using? If I'm not wildly off on that, then I'm thinking that we need the same form of protection for bounded staleness reads, even on their first (the "latest") attempt. If the first attempt is performed without any max_timestamp_bound at all, then how do we ensure that it doesn't dynamically choose a timestamp well in the future and beyond its descriptor's expiration?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/catalog/descs/leased_descriptors.go, line 59 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKZ8IVEIX0BDUSdkwj:-MhMbL8q-F1BxJoqK5hY:bpgvwhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/catalog/descs/leased_descriptors.go#L59)):*
> then how do we ensure that it doesn't dynamically choose a timestamp well in the future and beyond its descriptor's expiration?

I'm not as learned in the schema code, but we don't use deadline holder in this case `return nil` (it falls back to the existing leasing logic). see

```
func (tc *Collection) deadlineHolder(txn *kv.Txn) deadlineHolder {
	if tc.maxTimestampBoundDeadlineHolder.maxTimestampBound.IsEmpty() {
		return txn
	}
	return &tc.maxTimestampBoundDeadlineHolder
}
```


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/catalog/descs/leased_descriptors.go, line 59 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKZ8IVEIX0BDUSdkwj:-MhMbwm97wmPfaoX5AI7:byou4hq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/catalog/descs/leased_descriptors.go#L59)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

> then how do we ensure that it doesn't dynamically choose a timestamp well in the future and beyond its descriptor's expiration?

I'm not as learned in the schema code, but we don't use deadline holder in this case `return nil` (it falls back to the existing leasing logic). see

```
func (tc *Collection) deadlineHolder(txn *kv.Txn) deadlineHolder {
	if tc.maxTimestampBoundDeadlineHolder.maxTimestampBound.IsEmpty() {
		return txn
	}
	return &tc.maxTimestampBoundDeadlineHolder
}
```
</blockquote></details>

sorry, we don't use this deadline holder in that case (we fall back to the txn one), which updates the deadline.


<!-- Sent from Reviewable.io -->

otan committed:
kvfollowerreadsccl: implement txn retries for nearest_only=True

This implements the schema unavailability workaround for bounded
staleness reads, where if we had to bump the AOST min timestamp bound as
we fetched a schema descriptor which was more recent than the AOST min
timestamp bound. In this case, we look to retry, using an older schema
descriptor each time we go back.

We utilize the existing machinery around transaction retries to work.
Between retries, we keep track of a `MaxStalenessBound`, so that in
between retries we always read a lower bound.

There are a number of changes around retries needed to make this work:
* We preserve AOST between transaction retries. This is needed to
  prevent a case where the timestamp bound keeps progressing.
* with_min_timestamp / with_max_staleness now always return the context
  AS OF SYSTEM TIME. This is required as checks elsewhere require this
  to be equal to the evaluated time, but on txn retries.
* The schema descriptor code has to know the `MaxStalenessBound`. As
  such, we store this in the `Collection` object which is persisted
  inside `extraTxnState` and is preserved between retries.

Release note (sql change): Bounded staleness reads now retry
transactions when nearest_only=True and a schema change is detected
which may prevent a follower read from being served.

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68969)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @nvanbenschoten, and @otan)

___
*[pkg/sql/catalog/descs/leased_descriptors.go, line 59 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKZ8IVEIX0BDUSdkwj:-MhMci3R4ukpU__awjjj:b-1ke2k0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/catalog/descs/leased_descriptors.go#L59)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

sorry, we don't use this deadline holder in that case (we fall back to the txn one), which updates the deadline.
</blockquote></details>

Got it, I think that's ok, but then we need to do some extra work in `kv.Txn` to respect the deadline when `NegotiateAndSend` is used. That should be fine and I should have realized that's how this was going to work earlier. We'll just use the same `max_timestamp_bound` mechanism with some code in `NegotiateAndSend` like:
```go
if txn.mu.deadline != nil {
    ba.BoundedStaleness.MaxTimestampBound.Backward(*txn.mu.deadline)
}
```
I'll open an issue for this, though I won't get to it until next week.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/catalog/descs/leased_descriptors.go, line 59 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-MhKZ8IVEIX0BDUSdkwj:-MhMe_8U6rxG8ydB_zIg:b-fxdbp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e90277b8258d15f278aa01d90c78ce2f90daa7d6/pkg/sql/catalog/descs/leased_descriptors.go#L59)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Got it, I think that's ok, but then we need to do some extra work in `kv.Txn` to respect the deadline when `NegotiateAndSend` is used. That should be fine and I should have realized that's how this was going to work earlier. We'll just use the same `max_timestamp_bound` mechanism with some code in `NegotiateAndSend` like:
```go
if txn.mu.deadline != nil {
    ba.BoundedStaleness.MaxTimestampBound.Backward(*txn.mu.deadline)
}
```
I'll open an issue for this, though I won't get to it until next week.
</blockquote></details>

sounds good, yeah it smelt as thought it would have affected the nearest_only=False case as well.
if this LGTM and is the last comment i'm going to merge before you go :D


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68969#-:-MhMhmoG-A88ykIm_Tfs:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @otan)



<!-- Sent from Reviewable.io -->

otan commented:
cheers!

bors r=nvanbenschoten,ajwerner,rytaft
